README.md
#threejs演示文档

简单介绍three.js基础的功能组件

CHANGELOG.md
# Changelog

## v1.1.0

-创建本文档
-添加对three.js基础组件的介绍


js/slides.js
$(function() {
	$.deck('.slide .instant');

  var speed = 1;

  window.iframes = [];

  // Go to specific step
  function directorGo(iframe, step) {
    iframe.contentWindow && iframe.contentWindow.postMessage({ director: { args: [step] }}, '*');
  }

  // Pre-load and unload iframes one frame before/after
  var $iframes = {};
  $('.slide').each(function (i) {
    var $this = $(this);
    var $parents = $this.parents('.slide');
    var mask = $this.is('.instant') ? [i, i+1] : [i-1, i, i+1];

    // Build index of which iframes are active per slide
    if ($parents.length) $this = $parents;
    $.each(mask, function (i,v) {
      $iframes[v] = ($iframes[v] || $()).add($this.find('iframe'));
    });
  });

  function disable(iframe) {
    if (!$(iframe).data('src')) {
      var src = $(iframe).attr('src');
      $(iframe).data('src', src);
      iframe.onload = null;
      iframe.src = 'about:blank';

      iframes.splice(iframes.indexOf(iframe), 1);
    }
  }

  function enable(iframe, step) {
    var src = $(iframe).data('src');
    if (src) {
      iframe.onload = function () {
        iframe.onload = null;
        directorGo(iframe, step);
      }
      iframe.src = src;
      $(iframe).data('src', null);

      iframes.push(iframe);
    }
  }

  // Hide all iframes
  $('iframe').each(function () {
    disable(this);
  });

  // Respond to presentation deck navigation
  var $frames = null;

	$(document).bind('deck.change', function (e, from, to) {
    var out = [];

    $('#message').remove();

    function getTopSlide(step) {
      var $slide = $.deck('getSlide', step),
          $parents = $slide.parents('.slide');

      if ($parents.length) {
        $slide = $parents;
      }

      return $slide;
    }

    var $subslide = $.deck('getSlide', to);
    var $slide = getTopSlide(to);
    var step = $slide.find('.slide').index($subslide) + 2;

    // Sync up iframes to correct step
    $frames = $slide.find('iframe');
    $frames.each(function () {
      directorGo(this, step);
    });

    // Start playing videos
    $slide.find('video').each(function () {
      this.play();
    });
    $('.deck-container')[$slide.find('iframe.youtube').length ? 'addClass' : 'removeClass']('flat');

    // Stop old videos
    setTimeout(function () {
      $.deck('getSlide', from).find('video').each(function () {
        this.pause();
      });
    }, 500 / speed + 80);

    // Start at beginning or end of slides
    var go = to > from ? 1 : -1;

    // Pre-load iframes (but allow time for current transition)
    $iframes[to].each(function () {
      var iframe = this;
      setTimeout(function () { enable(iframe, go); }, 500 / speed + 80);
    });

    // Unload old iframes
    $('iframe').not($iframes[to]).each(function () {
      disable(this);
    });

    // Google Analytics
    if (window.gtag) {
      gtag('event', 'page_view', {
        // todo: this is the same for all 3 .html files
        'page_path': '/slide-' + to
      });
    }

  });

  if ( location.hash == "" ) {
    location.hash = "slide-0";
  }


  // xbox controller navigation

  window.addEventListener( 'vr controller connected', function( event ){

    var controller = event.detail;

    function next( event ){
      $.deck('next');
    }

    function prev( event ){
      $.deck('prev');
    }

    if ( controller.style == 'xbox' ) {

      controller.addEventListener( 'bumper-right press began', next );
      controller.addEventListener( 'trigger-right press began', next );
      controller.addEventListener( 'd-right press began', next );
      controller.addEventListener( 'thumbstick-left right press began', next );
      controller.addEventListener( 'thumbstick-right right press began', next );

      controller.addEventListener( 'bumper-left press began', prev );
      controller.addEventListener( 'trigger-left press began', prev );
      controller.addEventListener( 'd-left press began', prev );
      controller.addEventListener( 'thumbstick-left left press began', prev );
      controller.addEventListener( 'thumbstick-right left press began', prev );

    } else {

      controller.addEventListener( 'primary press began', next );

    }

  });

  loop();
  function loop() {
    requestAnimationFrame( loop );
    THREE.VRController.update();
  }

});

js/game/game.js
$(function () {
    initThree();
    initCamera();
    initScene();
    initLight();
    initObject();
    initListener();
    initStats();
    initGUI();
    animate();
});

var controls = new function () {
    this.飞机螺旋桨速度 = 0.25;
    this.背景云朵速度 = 0.01;
    this.光源强度 = 0.9;
    //......
};

//渲染器
var renderer;
var width, height;
function initThree() {
    width = window.innerWidth;
    height = window.innerHeight;
    renderer = new THREE.WebGLRenderer({
        // 在 css 中设置背景色透明显示渐变色
        alpha: true,
        // 开启抗锯齿，但这样会降低性能。
        antialias: true
    });
    // 打开渲染器的阴影地图
    renderer.shadowMap.enabled = true;
    //设置像素比例
    renderer.setPixelRatio(window.devicePixelRatio);
    // 定义渲染器的尺寸；在这里它会填满整个屏幕
    renderer.setSize(window.innerWidth, window.innerHeight);
    document.body.appendChild(renderer.domElement);
}

//相机
var camera;
function initCamera() {
    aspectRatio = width / height;
    fieldOfView = 60;
    nearPlane = 1;
    farPlane = 10000;

    /**
     * PerspectiveCamera 透视相机
     * @param fieldOfView 视角
     * @param aspectRatio 纵横比
     * @param nearPlane 近平面
     * @param farPlane 远平面
     */
    camera = new THREE.PerspectiveCamera(
        fieldOfView,
        aspectRatio,
        nearPlane,
        farPlane
    );
    camera.position.x = 0;
    camera.position.y = 100;
    camera.position.z = 200;
}

//场景
var scene;
function initScene() {
    scene = new THREE.Scene();
    // 在场景中添加雾的效果；样式上使用和背景一样的颜色
    scene.fog = new THREE.Fog(0xf7d9aa, 100, 950);
}

//光照
var hemisphereLight, directionalLight, ambientLight;
function initLight() {
    // 渐变的半球光；
    // 第一个参数是天空的颜色，第二个参数是地上的颜色，第三个参数是光源的强度
    hemisphereLight = new THREE.HemisphereLight(0xaaaaaa, 0x000000, controls.光源强度);
    scene.add(hemisphereLight);

    // 从一个特定的方向的照射的平行光
    // 第一个参数是关系颜色，第二个参数是光源强度
    directionalLight = new THREE.DirectionalLight(0xffffff, controls.光源强度);
    //设置光源方向
    directionalLight.position.set(150, 350, 350);
    //开启光源阴影
    directionalLight.castShadow = true;
    //定义可见区域的投射阴影
    directionalLight.shadow.camera.left = -400;
    directionalLight.shadow.camera.right = 400;
    directionalLight.shadow.camera.top = 400;
    directionalLight.shadow.camera.bottom = -400;
    directionalLight.shadow.camera.near = 1;
    directionalLight.shadow.camera.far = 1000;
    //定义阴影分辨率
    directionalLight.shadow.mapSize.width = 2048;
    directionalLight.shadow.mapSize.height = 2048;
    scene.add(directionalLight);

    //添加环境光，设置颜色、光源强度
    ambientLight = new THREE.AmbientLight(0xdc8874, controls.光源强度*0.6);
    scene.add(ambientLight);
}

//自定义物体
var sea;
var sky;
var airPlane;
function initObject() {
    //建造大海
    sea = new Sea();
    sea.mesh.position.y = -600;
    scene.add(sea.mesh);

    //建造天空
    sky = new Sky();
    sky.mesh.position.y = -600;
    scene.add(sky.mesh);

    //建造飞机
    airPlane = new AirPlane();
    airPlane.mesh.scale.set(0.25, 0.25, 0.25);
    airPlane.mesh.position.y = 100;
    scene.add(airPlane.mesh);

}

//动画
function animate() {
    render();
    requestAnimationFrame(animate);

    stats.update();
}

//渲染
function render() {
    updateLight();
    updateAirPlane();
    airPlane.pilot.updateHairs();
    updateCameraFov();
    sea.moveWaves();
    sky.mesh.rotation.z += controls.背景云朵速度;
    renderer.render(scene, camera);
}

function updateLight(){
    hemisphereLight.intensity = controls.光源强度;
    directionalLight.intensity = controls.光源强度;
    ambientLight.intensity = controls.光源强度*0.6;
}

function updateAirPlane(){
    var x = normalize(mouseX, -0.75, 0.75, -100, 100);
    var y = normalize(mouseY, -0.75, 0.75, 25, 175);

    airPlane.mesh.position.y += (y - airPlane.mesh.position.y)*0.1;
    airPlane.mesh.rotation.z = (y - airPlane.mesh.position.y)*0.0128;
    airPlane.mesh.rotation.x = (airPlane.mesh.position.y - y)*0.0064;
    airPlane.propeller.rotation.x += controls.飞机螺旋桨速度;
}

function updateCameraFov(){
    camera.fov = normalize(mouseX,-1,1,50, 70);
    camera.updateProjectionMatrix();
}

function normalize(v,vmin,vmax,tmin, tmax){
    var nv = Math.max(Math.min(v,vmax), vmin);
    var dv = vmax-vmin;
    var pc = (nv-vmin)/dv;
    var dt = tmax-tmin;
    var tv = tmin + (pc*dt);
    return tv; }

//性能监听器
var stats;
function initStats() {
    stats = new Stats();
    stats.domElement.style.position = 'absolute';
    stats.domElement.style.top = '0px';
    document.body.appendChild(stats.domElement);
}

//参数调节工具
function initGUI() {
    var gui = new dat.GUI();
    gui.add(controls, '飞机螺旋桨速度', 0, 0.5);
    gui.add(controls, '背景云朵速度', 0, 0.1);
    gui.add(controls, '光源强度', 0, 1.0);
}

//监听器
function initListener() {
    // 监听屏幕，缩放屏幕更新相机和渲染器的尺寸
    window.addEventListener('resize', handleWindowResize, false);
    //监听鼠标移动
    document.addEventListener('mousemove', handleMouseMove, false);
}

function handleWindowResize() {
    // 更新渲染器的宽度和高度以及相机的纵横比
    width = window.innerWidth;
    height = window.innerHeight;
    renderer.setSize(width, height);
    camera.aspect = width / height;
    camera.updateProjectionMatrix();
}

var mouseX = 0, mouseY = 0;
function handleMouseMove(event) {
    var clientX = -1 + (event.clientX / width) * 2;
    var clientY = 1 - (event.clientY / height) * 2;
    mouseX = clientX;
    mouseY = clientY;
}


js/game/game-obj.js
//色彩
var Colors = {
    red: 0xf25346,
    white: 0xd8d0d1,
    brown: 0x59332e,
    pink: 0xF5986E,
    brownDark: 0x23190f,
    blue: 0x68c3c0,
    yellow:0xf4ce93
};

//大海
Sea = function() {
    //创建一个圆柱体
    // 参数为：顶面半径，底面半径，高度，半径分段，高度分段
    var geometry = new THREE.CylinderGeometry(600, 600, 800, 40, 10);
    //旋转圆柱体
    geometry.applyMatrix(new THREE.Matrix4().makeRotationX(-Math.PI / 2));

    //合并圆柱体顶点
    geometry.mergeVertices();

    this.waves = [];

    for (var i = 0; i < geometry.vertices.length; i++){
        var vertice = geometry.vertices[i];
        this.waves.push({
            x: vertice.x,
            y: vertice.y,
            z: vertice.z,
            ang: Math.random()*Math.PI*2, //角度
            amp: 5 + Math.random()*15, //距离
            speed: 0.016 + Math.random()*0.032 //速度（0.016~0.048度/帧）
        });
    }

    //创建Phong材质
    var material = new THREE.MeshPhongMaterial({
        color: Colors.blue,
        transparent: true, //是否透明
        opacity: 0.8, //不透明度
        flatShading: true
    });
    //组合圆柱体和材质
    this.mesh = new THREE.Mesh(geometry, material);
    //设置允许接受阴影
    this.mesh.receiveShadow = true;
}

//海浪
Sea.prototype.moveWaves = function(){
    var vertices = this.mesh.geometry.vertices;
    for (var i = 0; i < vertices.length; i++){
        var vertice = vertices[i];
        var wave = this.waves[i];

        vertice.x = wave.x + Math.cos(wave.ang) * wave.amp;
        vertice.y = wave.y + Math.sin(wave.ang) * wave.amp;

        wave.ang += wave.speed;
    }

    //告诉渲染器代表大海的几何体发生改变
    this.mesh.geometry.verticesNeedUpdate = true;
    sea.mesh.rotation.z += .005;
}

//云朵
Cloud = function () {
    // 创建一个空的容器放置不同形状的云
    this.mesh = new THREE.Object3D();

    //创建正方体
    var geometry = new THREE.BoxGeometry(20, 20, 20);

    //创建材质
    var material = new THREE.MeshPhongMaterial({
        color: Colors.white,
    });

    // 随机复制正方体
    var nRandom = 3 + Math.floor(Math.random() * 3);
    for (var i = 0; i < nRandom; i++) {
        var mesh = new THREE.Mesh(geometry, material);
        // 随机设置每个正方体的位置和旋转角度
        mesh.position.x = i * 15;
        mesh.position.y = Math.random() * 10;
        mesh.position.z = Math.random() * 10;
        mesh.rotation.z = Math.random() * Math.PI * 2;
        mesh.rotation.y = Math.random() * Math.PI * 2;
        // 随机设置正方体的大小
        var s = 0.1 + Math.random() * 0.9;
        mesh.scale.set(s, s, s);
        // 允许每个正方体生成投影和接收阴影
        mesh.castShadow = true;
        mesh.receiveShadow = true;
        this.mesh.add(mesh);
    }
}

//天空
Sky = function () {
    // 创建一个空的容器
    this.mesh = new THREE.Object3D();

    this.nClouds = 20;

    //把云根据弧度均匀放置
    var angle = Math.PI * 2 / this.nClouds;
    for (var i = 0; i < this.nClouds; i++) {
        var cloud = new Cloud();
        // 设置每朵云的旋转角度和位置
        var anglei = angle * i;
        //这是云的最终角度
        var h = 750 + Math.random() * 200;
        // 把极坐标转换成笛卡坐标
        cloud.mesh.position.y = Math.sin(anglei) * h;
        cloud.mesh.position.x = Math.cos(anglei) * h;
        // 根据云的位置旋转它
        cloud.mesh.rotation.z = anglei + Math.PI / 2;
        //把云放置在场景中的随机深度位置
        cloud.mesh.position.z = -400 - Math.random() * 400;
        // 设置每朵云的随机大小
        var s = 1 + Math.random() * 2;
        cloud.mesh.scale.set(s, s, s);

        this.mesh.add(cloud.mesh);

    }
}

//驾驶员
Pilot = function(){
    this.mesh = new THREE.Object3D();
    this.angleHairs=0;

    //身体
    var geometryBody = new THREE.BoxGeometry(15,15,15);
    var materialBody = new THREE.MeshPhongMaterial({
        color:Colors.brown,
        shading:THREE.FlatShading
    });
    var body = new THREE.Mesh(geometryBody, materialBody);
    body.position.set(2,-12,0);
    this.mesh.add(body);

    //脸
    var geometryFace = new THREE.BoxGeometry(10,10,10);
    var materialFace = new THREE.MeshLambertMaterial({
        color:Colors.pink
    });
    var face = new THREE.Mesh(geometryFace, materialFace);
    this.mesh.add(face);

    //头发
    var hairs = new THREE.Object3D();

    var geometryHair = new THREE.BoxGeometry(4,4,4);
    var materialHair = new THREE.MeshLambertMaterial({
        color:Colors.brown
    });
    var hair = new THREE.Mesh(geometryHair, materialHair);
    hair.geometry.applyMatrix(new THREE.Matrix4().makeTranslation(0,2,0));
    //顶部头发
    this.hairsTop = new THREE.Object3D();
    for (var i=0; i<12; i++){
        var h = hair.clone();
        var col = i%3;
        var row = Math.floor(i/3);
        var startPosZ = -4;
        var startPosX = -4;
        h.position.set(startPosX + row*4, 0, startPosZ + col*4);
        this.hairsTop.add(h);
    }
    hairs.add(this.hairsTop);

    //左右两侧头发
    var geometryHairSide = new THREE.BoxGeometry(12,4,2);
    geometryHairSide.applyMatrix(new THREE.Matrix4().makeTranslation(-6,0,0));
    var hairSideRight = new THREE.Mesh(geometryHairSide, materialHair);
    hairSideRight.position.set(8,-2,6);
    hairs.add(hairSideRight);
    var hairSideLeft = hairSideRight.clone();
    hairSideLeft.position.set(8,-2,-6);
    hairs.add(hairSideLeft);

    //后侧头发
    var geometryHairBack = new THREE.BoxGeometry(2,8,10);
    var hairBack = new THREE.Mesh(geometryHairBack, materialHair);
    hairBack.position.set(-1,-4,0)
    hairs.add(hairBack);
    hairs.position.set(-5,5,0);
    this.mesh.add(hairs);

    //眼镜镜面
    var geometryGlass = new THREE.BoxGeometry(5,5,5);
    var materialGlass = new THREE.MeshLambertMaterial({color:Colors.brown});
    var glassRight = new THREE.Mesh(geometryGlass,materialGlass);
    glassRight.position.set(6,0,3);
    this.mesh.add(glassRight);
    var glassLeft = glassRight.clone();
    glassLeft.position.z = -glassRight.position.z
    this.mesh.add(glassLeft);

    //眼镜腿
    var geometryGlassLeg = new THREE.BoxGeometry(11,1,11);
    var glassLeg = new THREE.Mesh(geometryGlassLeg, materialGlass);
    this.mesh.add(glassLeg);

    //耳朵
    var geometryEar = new THREE.BoxGeometry(2,3,2);
    var earRight = new THREE.Mesh(geometryEar,materialFace);
    earRight.position.set(0,0,-6);
    this.mesh.add(earRight);
    var earLeft = earRight.clone();
    earLeft.position.set(0,0,6);
    this.mesh.add(earLeft);
}

//头发顶部的飘动
Pilot.prototype.updateHairs = function(){
    var hairs = this.hairsTop.children;
    for (var i=0; i<hairs.length; i++){
        var h = hairs[i];
        h.scale.y = .75 + Math.cos(this.angleHairs+i/3)*.25;
    }
    this.angleHairs += 0.16;
}

//飞机
AirPlane = function () {
    this.mesh = new THREE.Object3D();

    //创建机舱
    var geometryCockpit = new THREE.BoxGeometry(80, 40, 40);
    // 移动长方形顶点的 x, y, z 属性
    geometryCockpit.vertices[4].y-=10;
    geometryCockpit.vertices[4].z+=20;
    geometryCockpit.vertices[5].y-=10;
    geometryCockpit.vertices[5].z-=20;
    geometryCockpit.vertices[6].y+=30;
    geometryCockpit.vertices[6].z+=20;
    geometryCockpit.vertices[7].y+=30;
    geometryCockpit.vertices[7].z-=20;

    var materialCockpit = new THREE.MeshPhongMaterial({
        color: Colors.red,
        flatShading: true
    });

    var cockpet = new THREE.Mesh(geometryCockpit, materialCockpit);
    cockpet.castShadow = true;
    cockpet.receiveShadow = true;
    this.mesh.add(cockpet);

    //创建引擎
    var geometryEngine = new THREE.BoxGeometry(20, 40, 40);
    var materialEngine = new THREE.MeshPhongMaterial({
        color: Colors.white,
        flatShading: true
    });
    var engine = new THREE.Mesh(geometryEngine, materialEngine);
    engine.castShadow = true;
    engine.receiveShadow = true;
    engine.position.x = 50;
    this.mesh.add(engine);

    //创建机尾
    var geometryTailPlane = new THREE.BoxGeometry(15, 20, 5);
    var materialTailPlane = new THREE.MeshPhongMaterial({
        color: Colors.red,
        flatShading: true
    });
    var tailPlane = new THREE.Mesh(geometryTailPlane, materialTailPlane);
    tailPlane.castShadow = true;
    tailPlane.receiveShadow = true;
    tailPlane.position.set(-40, 20, -5);
    this.mesh.add(tailPlane);

    //创建机翼
    var geometrySideWing = new THREE.BoxGeometry(30, 5, 120);
    var materialSideWing = new THREE.MeshPhongMaterial({
        color: Colors.red,
        flatShading: true
    });
    var sideWing = new THREE.Mesh(geometrySideWing, materialSideWing);
    sideWing.castShadow = true;
    sideWing.receiveShadow = true;
    this.mesh.add(sideWing);

    //创建挡风玻璃
    var geometryWindshield = new THREE.BoxGeometry(3,15,20);
    var materialWindshield = new THREE.MeshPhongMaterial({
        color:Colors.white,
        transparent:true,
        opacity:.3,
        flatShading: true
    });
    var windshield = new THREE.Mesh(geometryWindshield, materialWindshield);
    windshield.position.set(5,27,0);
    windshield.castShadow = true;
    windshield.receiveShadow = true;
    this.mesh.add(windshield);

    //创建螺旋桨
    var geometryPropeller = new THREE.BoxGeometry(20, 10, 10);
    geometryPropeller.vertices[4].y-=5;
    geometryPropeller.vertices[4].z+=5;
    geometryPropeller.vertices[5].y-=5;
    geometryPropeller.vertices[5].z-=5;
    geometryPropeller.vertices[6].y+=5;
    geometryPropeller.vertices[6].z+=5;
    geometryPropeller.vertices[7].y+=5;
    geometryPropeller.vertices[7].z-=5;
    var materialPropeller = new THREE.MeshPhongMaterial({
        color: Colors.brown,
        flatShading: true
    });
    this.propeller = new THREE.Mesh(geometryPropeller, materialPropeller);
    this.propeller.castShadow = true;
    this.propeller.receiveShadow = true;

    //创建螺旋桨桨叶
    var geometryBlade = new THREE.BoxGeometry(1, 80, 10);
    var materialBlade = new THREE.MeshPhongMaterial({
        color: Colors.brownDark,
        flatShading: true
    });
    var bladeRight = new THREE.Mesh(geometryBlade, materialBlade);
    bladeRight.castShadow = true;
    bladeRight.receiveShadow = true;
    bladeRight.position.set(8, 0, 0);

    var bladeLeft = bladeRight.clone();
    bladeLeft.castShadow = true;
    bladeLeft.receiveShadow = true;
    bladeLeft.rotation.x = Math.PI/2;

    //组合螺旋桨和桨叶
    this.propeller.add(bladeRight);
    this.propeller.add(bladeLeft);
    this.propeller.position.set(60, 0, 0);
    this.mesh.add(this.propeller);

    //创建轮子
    //防护罩
    var geometryWheelShield = new THREE.BoxGeometry(30,15,10);
    var materialWheelShield = new THREE.MeshPhongMaterial({
        color:Colors.red,
        flatShading: true
    });
    var wheeleelShieldRight = new THREE.Mesh(geometryWheelShield,materialWheelShield);
    wheeleelShieldRight.position.set(25,-18,25);
    this.mesh.add(wheeleelShieldRight);

    //轮胎
    var geometryWheelTire = new THREE.CylinderGeometry(12, 12, 4, 40);
    geometryWheelTire.applyMatrix(new THREE.Matrix4().makeRotationX(-Math.PI / 2));
    var materialWheelTire = new THREE.MeshPhongMaterial({
        color:Colors.brownDark,
        flatShading: true
    });
    var wheelTireRight = new THREE.Mesh(geometryWheelTire,materialWheelTire);
    wheelTireRight.position.set(25,-28,25);

    //轮轴
    var geometryWheelAxis = new THREE.CylinderGeometry(5, 5, 6, 40);
    geometryWheelAxis.applyMatrix(new THREE.Matrix4().makeRotationX(-Math.PI / 2));
    var materialWheelAxis = new THREE.MeshPhongMaterial({
        color:Colors.brown,
        flatShading: true
    });
    var wheelAxis = new THREE.Mesh(geometryWheelAxis,materialWheelAxis);
    wheelTireRight.add(wheelAxis);

    this.mesh.add(wheelTireRight);

    var wheeleelShieldLeft = wheeleelShieldRight.clone();
    wheeleelShieldLeft.position.z = -wheeleelShieldRight.position.z ;
    this.mesh.add(wheeleelShieldLeft);

    var wheelTireLeft = wheelTireRight.clone();
    wheelTireLeft.position.z = -wheelTireRight.position.z;
    this.mesh.add(wheelTireLeft);

    var wheelTireRear = wheelTireRight.clone();
    wheelTireRear.scale.set(0.5,0.5,0.5);
    wheelTireRear.position.set(-35,-5,0);
    this.mesh.add(wheelTireRear);

    //后轮悬架
    var geometryRearWheelSuspension = new THREE.BoxGeometry(4,20,4);
    //旋转
    geometryRearWheelSuspension.applyMatrix(new THREE.Matrix4().makeTranslation(0,10,0))
    var materialRearWheelSuspension = new THREE.MeshPhongMaterial({
        color:Colors.red,
        flatShading: true
    });
    var rearWheelSuspension = new THREE.Mesh(geometryRearWheelSuspension,materialRearWheelSuspension);
    rearWheelSuspension.position.set(-35,-5,0);
    rearWheelSuspension.rotation.z = -.3;
    this.mesh.add(rearWheelSuspension);

    //驾驶员入座
    this.pilot = new Pilot();
    this.pilot.mesh.position.set(-10,27,0);
    this.mesh.add(this.pilot.mesh);

    this.mesh.castShadow = true;
    this.mesh.receiveShadow = true;
}

